### PR TITLE
add new phase  to WasmPlugin

### PR DESCRIPTION
**Pull request by zirain**
xref: https://github.com/istio/istio/issues/50012

this will allow user insert WasmPlugin at the start of the filter chain.

### CODE REVIEW

File: extensions/v1alpha1/wasm.proto [hunk 1]
```diff
@@ -414,6 +414,9 @@ enum PluginPhase {
 
   // Insert plugin before Istio stats filters and after Istio authorization filters.
   STATS = 3;
+
+  // Insert plugin before Istio metadata filters, This will generally be at the start of the filter chain.
+  METADATA_EXCHANGE = 4;
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply extensions/v1alpha1/wasm.proto:419]: MX filter should not be assumed to exist forever, and moreover, the data it provides cannot be trusted.
Maybe it's better to configure MX filter not to strip the header instead?

[the remark this answers is not in the document]
Comment (zirain) [reply extensions/v1alpha1/wasm.proto:419]: sg

[the remark this answers is not in the document]
Comment (zirain) [reply extensions/v1alpha1/wasm.proto]: rethink about this, still worth add a new phase enable user insert WasmPlugin at the start of the filter chain?

**Review by ericvn — APPROVED** [review]

**Comment by kyessenov** [comment]
We deliberately tried to assign semantic meaning to each phase. I don't know what "initial" actually means.

**Comment by howardjohn** [comment]
Sorry didn't realize there was open discussion here before merging. @zirain can you discuss with @kyessenov and make any needed changes/revert before 1.22 if needed

**Comment by kyessenov** [comment]
The config should be about the intent, not the mechanics. If you can explain what "initial" is supposed to be used for, and rename it, that would be enough.

**Comment by ericvn** [comment]
To further expand on this, what filters are run before the Istio authentication filters? From the issue, one would be the istio.metadata_exchange filter. What phase would include that filter and other filters that one would want to put their filter before? It sounds like calling that phase initial may not be the best. My read was that using Initial just meant that the filters would be put at the front of the line. I don't have a better name in mind.

**Comment by zirain** [comment]
`INITIAL` means `Insert plugin at the start of the filter chain.`, just the other side of `UNSPECIFIED_PHASE`(`This will generally be at the end of the filter chain, right before the Router.`)

I think maybe `METADATA_EXCHANGE` is a better option, but as you said before istio should not assume it will be always there, do you have a better proposal than `INITIAL `?

**Comment by ramaraochavali** [comment]
IMO, AUTHN phase should cover most of the cases that need to run. I do not know what would be the use case for being the filter first? METADATA_EXCHANGE seems like a special case - may be we should the change the implementation of AUTHN Phase to insert before Metadata Exchange ?

**Comment by hzxuzhonghu** [comment]
Something like iptabels raw table? Allow any audit, any data preprocessing?

**Comment by kyessenov** [comment]
@hzxuzhonghu Isn't that authentication? If you want check or assert some security property, that's authentication?

**Comment by bleggett** [comment]
April 10 WG decision was to revert this change, and solve the metadata exchange UX issue here without adding a new WasmPlugin stage: https://docs.google.com/document/d/1wsa06GGiq1LEGwhkiPP0FKIZJqdAiue-VeBonWAzAyk/edit#heading=h.6fcb5py1h0hy

**Comment by zirain** [comment]
> solve the metadata exchange UX 

Can you explain how to do that?

**Comment by kyessenov** [comment]
@zirain Can you read the metadata from the filter state? The details for encoding in either the filter state or the header are subject to change, so either way it's not a fully supported configuration.

**Pull request merged**